common/wallet_tx.c

```diff
@@ -104,3 +165,62 @@ struct command_result *wtx_select_utxos(struct wallet_tx *tx,
 	}
 	return NULL;
 }
+
+struct command_result *wtx_from_utxos(struct wallet_tx *tx,
+					u32 fee_rate_per_kw,
+					size_t out_len,
+					u32 maxheight,
+					const struct utxo **utxos)
+{
+	struct command_result *res;
+	size_t weight, i;
+	struct amount_sat total_amount, fee_estimate;
+
+	tx->change = AMOUNT_SAT(0);
+	tx->change_key_index = 0;
+	total_amount = AMOUNT_SAT(0);
+
+	/* The transaction has `tal_count(tx.utxos)` inputs and one output */
+	/* (version + in count + out count + locktime)  (index + value + script length) */
+	weight = 4 * (4 + 1 + 1 + 4) + 4 * (8 + 1 + out_len);
+	for (i = 0; i < tal_count(utxos); i++) {
+        if (!*utxos[i]->blockheight || *utxos[i]->blockheight > maxheight) {
+			tal_arr_remove(&utxos, i);
+			continue;
+		}
+		/* txid + index + sequence + script_len */
+		weight += (32 + 4 + 4 + 1) * 4;
+		/* P2SH variants include push of <0 <20-byte-key-hash>> */
+		if (utxos[i]->is_p2sh)
+			weight += 23 * 4;
+		/* Account for witness (1 byte count + sig + key) */
+		weight += 1 + (1 + 73 + 1 + 33);
+		if (!amount_sat_add(&total_amount, total_amount, utxos[i]->amount))
+			fatal("Overflow when computing input amount");
+	}
+	tx->utxos = utxos;
+
+	if (!tx->all_funds && amount_sat_less(tx->amount, total_amount)
+			&& !amount_sat_sub(&tx->change, total_amount, tx->amount))
+		fatal("Overflow when computing change");
+	if (amount_sat_greater_eq(tx->change, get_chainparams(tx->cmd->ld)->dust_limit)) {
+		tx->change_key_index = wallet_get_newindex(tx->cmd->ld);
+		/* Add the change output's weight */
+		weight += (8 + 1 + out_len) * 4;
+	}
+
+	fee_estimate = amount_tx_fee(fee_rate_per_kw, weight);
+
+	if (tx->all_funds || amount_sat_eq(tx->change, AMOUNT_SAT(0))) {
+		tx->amount = total_amount;
+        if (!amount_sat_sub(&tx->amount, tx->amount, fee_estimate))
+			return command_fail(tx->cmd, FUND_CANNOT_AFFORD,
+								"Cannot afford transaction");
+	} else {
+        if (!amount_sat_sub(&tx->change, tx->change, fee_estimate))
+            return command_fail(tx->cmd, FUND_CANNOT_AFFORD,
+								"Cannot afford transaction");
+	}
+	res = check_amount(tx, tx->amount);
+    return res;
```

Comment:
indent is weird, but just `return check_amount...` instead of a temporary variable?